driver-core/src/main/com/mongodb/client/model/expressions/ArrayExpression.java

```diff
@@ -24,4 +27,38 @@
  */
 public interface ArrayExpression<T extends Expression> extends Expression {
 
+    /**
+     * Returns an array consisting of those elements in this array that match
+     * the given predicate condition. Evaluates each expression in this array
+     * according to the cond function. If cond evaluates to logical true, then
+     * the element is preserved; if cond evaluates to logical false, the element
+     * is omitted.
+     *
+     * @param cond the function to apply to each element
+     * @return the new array
+     */
+    ArrayExpression<T> filter(Function<T, BooleanExpression> cond);
+
+    /**
+     * Returns an array consisting of the results of applying the given function
+     * to the elements of this array.
+     *
+     * @param in the function to apply to each element
+     * @return the new array
+     * @param <R> the type contained in the resulting array
+     */
+    <R extends Expression> ArrayExpression<R> map(Function<T, ? extends R> in);
+
+    /**
+     * Performs a reduction on the elements of this array, using the provided
+     * identity value and an associative accumulation function, and returns
+     * the reduced value. The initial value must be the identity value for the
+     * reducing function.
+     *
+     * @param initialValue the identity for the reducing function
+     * @param in the associative accumulation function
+     * @return the reduced value
+     */
+    T reduce(T initialValue, BiFunction<T,  T,  T> in);
```

Comment:
Can you think of a way to express multiple reductions, so we can support something like:

```
{
    $reduce: {
       input: ["1", "2", "3"],
       initialValue: {sum: 0, product: 0, concat: ""},
       in: { 
          sum :    { $sum :      [{$toInt: "$$this"}, "$$value.sum"] },
          product: { $multiply : [{$toInt: "$$this"}, "$$value.product"] }
          concat:  { $concat   : ["$$this",           "$$value.concat"] }
        }   
     }
 }
```

?

Reply:
This can be accomplished with the existing method:

```
assertExpression(BsonDocument.parse("{'sum': 10, 'product': 24, 'concat': '1234'}"),

ofStringArray("1", "2", "3", "4")
        // map:
        .<DocumentExpression>map(v -> ofEmptyDocument()
                .setField("sum", v.toInt())
                .setField("product", v.toInt())
                .setField("concat", v))
        // identity:
        .reduce(ofEmptyDocument()
                        .setField("sum", of(0))
                        .setField("product", of(1))
                        .setField("concat", of("")),
        // reduction:
        (a, b) -> ofEmptyDocument()
                .setField("sum", a.getFieldInteger("sum").add(b.getFieldInteger("sum")))
                .setField("product", a.getFieldInteger("product").multiply(b.getFieldInteger("product")))
                .setField("concat", a.getFieldString("concat").concat(b.getFieldString("concat")))));
```

Above test passes. However, I am unsure if the server knows how to move a map into a reduction, or whether that might affect performance.

Compact with a schema object:

```
ofStringArray("1", "2", "3", "4")
        .map(v -> new Result(v.toInt(), v.toInt(), v))
        .reduce(new Result(0, 1, ""), (a, b) -> new Result(
                a.getSum().add(b.getSum()),
                a.getProduct().multiply(b.getProduct()),
                a.getConcat().concat(b.getConcat())) ))
```

Could try to push it to something approximating:

```
ofStringArray("1", "2", "3", "4").reduce(v -> result()
        .set("sum", v -> v.toInt().add())
        .set("product", v -> v.toInt().multiply())
        .set("concat", v -> v.concat()))
```

I think I'd need to have a closer look at $group. I wonder if this last variant (which is basically syntax sugar) obscures what is actually going on (a map followed by a reduce). On the other hand, I like that this last variant hides the identity value.

Reply:
What MQL does that first bit of code ("with the existing method") generate?

Reply:
```
{
  "$reduce":{
    "input":{
      "$map":{
        "input":[
          "1",
          "2",
          "3",
          "4"
        ],
        "in":{
          "$setField":{
            "field":"concat",
            "input":{
              "$setField":{
                "field":"product",
                "input":{
                  "$setField":{
                    "field":"sum",
                    "input":{
                      "$literal":{
                        
                      }
                    },
                    "value":{
                      "$toInt":"$$this"
                    }
                  }
                },
                "value":{
                  "$toInt":"$$this"
                }
              }
            },
            "value":"$$this"
          }
        }
      }
    },
    "initialValue":{
      "$setField":{
        "field":"concat",
        "input":{
          "$setField":{
            "field":"product",
            "input":{
              "$setField":{
                "field":"sum",
                "input":{
                  "$literal":{
                    
                  }
                },
                "value":0
              }
            },
            "value":1
          }
        },
        "value":""
      }
    },
    "in":{
      "$setField":{
        "field":"concat",
        "input":{
          "$setField":{
            "field":"product",
            "input":{
              "$setField":{
                "field":"sum",
                "input":{
                  "$literal":{
                    
                  }
                },
                "value":{
                  "$add":[
                    {
                      "$getField":{
                        "input":"$$value",
                        "field":"sum"
                      }
                    },
                    {
                      "$getField":{
                        "input":"$$this",
                        "field":"sum"
                      }
                    }
                  ]
                }
              }
            },
            "value":{
              "$multiply":[
                {
                  "$getField":{
                    "input":"$$value",
                    "field":"product"
                  }
                },
                {
                  "$getField":{
                    "input":"$$this",
                    "field":"product"
                  }
                }
              ]
            }
          }
        },
        "value":{
          "$concat":[
            {
              "$getField":{
                "input":"$$value",
                "field":"concat"
              }
            },
            {
              "$getField":{
                "input":"$$this",
                "field":"concat"
              }
            }
          ]
        }
      }
    }
  }
}
```